CBVCorrector working with 20-second CBVs.

jcsmithhere opened the pull request:
Here is a branch to get the CBVCorrector working with 20-second "fast" CBVs.

CBVCorrector will now automatically identify if the TESS light curve under study is 20-second or 2-minute cadence data and download the appropriate CBVs. The tutorials have been updated to explain the new functionality. 

For Kepler, only 30-minute CBVs are available so not much changed there.

New meta fields have been created in the light curve objects to help identify the cadence type:
```
lc.meta["EXPTIME"] = lc.meta.get("FRAMETIM") * lc.meta.get("NUM_FRM")
lc.meta["CADENCE_TYPE"] = (TESS: 'long', 'short', 'fast) or (KEPLER/K2: 'ffi', 'long', 'short')
```
Helper functions created are `io/tess.get_tess_cadence_type` and `io/kepler.get_kepler_cadence_type` to determine the cadence type from the exposure time. A third helper function is `cbvcorrector._determine_exptime_from_lc` but that one is not intended for public use.

It would be appreciated if someone could go through the tutorials and see if everything works for them and makes sense.

barentsen commented:
Thanks Jeff -- I expect to test/review this first thing tomorrow!

jcsmithhere commented:
Thanks, @barentsen! I also asked a couple other people to look at it.

barentsen commented:
> I also asked a couple other people to look at it.

Perfect.  The more feedback the better!!

barentsen commented on src/lightkurve/io/tess.py:
```diff
@@ -46,4 +46,48 @@ def read_tess_lightcurve(
+    The options are:
+    if   exptime < 60   : "fast"  (i.e. 20-second)
+    elif exptime < 300  : "short" (i.e. 2-minute)
+    else                : "long"  (i.e. 30 or 10 minute FFI)
```
I wonder if we want to use the string `"ffi"` instead of `"long"`, because I often hear people say "TESS FFI light curves" and I've ~never heard "TESS long cadence light curves".  (But I might be wrong -- is "long" is already adopted somewhere?)

jcsmithhere replied:
I'm debating this too. I've heard both. The mission uses FFI but I've seen others say "long". I would prefer "ffi'. If you are good with that then I'll change it to 'ffi'

barentsen replied:
I think "ffi" is the way to go!

jcsmithhere replied:
Terminology changed to 'ffi' and merged to branch.

dacmess commented:
@jcsmithhere I'm am but an egg and am not sure of the correct way to comment on your pull request, but I think for general comments, I can just put them here(?). 
In Tutorial 2-2-how-to-use-cbvs:  This tutorial worked fine for me and was clear.
-In the intro, you might change "Single-Scale" and "Multi-Scale" to "SingleScale" and "MultiScale" to match the actual python keyword values for cbv_type.
-You mention setting extrapolate=False sets the CBV values to 0 and will likely lead to a bad fit, but you might caution about extrapolate=True also likely being problematic and maybe it's better to crop the LC (I realize this was in the original code and not just introduced).

dacmess commented on docs/source/tutorials/2-creating-light-curves/2-3-how-to-use-cbvcorrector.ipynb:
```diff
@@ -146,9 +166,9 @@
+    "- **Under-fitting metric**: Measures the mean residual target to target Pearson correlation between the target under study and a selection of neighboring targets. This metric will find and download a selection of neighboring SPOC SAP targets in RA and Decl. until a minimum number is found. It attempts to find lightcurves with the same cadence type as the target under study. The metric is cal ...
     "\n",
-    "_The Goodness Metrics are not perfect!_ They are an estimate of over- and under-fitting and are to be used as a guideline along other other metrics to assess the quality of your light curve."
+    "_The Goodness Metrics are not perfect!_ They are an estimate of over- and under-fitting and are to be used as a guideline along other metrics to assess the quality of your light curve."
```
"...used as a guideline along with other metrics...."

jcsmithhere replied:
Fixed

dacmess commented:
On tutorial 2-3-how-to-use-cbvcorrector:
typo in description after correct_gaussian_prior() cell: should be "The very strong dips right at the beginning..." rather than "lips"
More generally, my correction for TIC 99180739 doesn't look like a good correction, am I missing something?
<img width="724" alt="Screen Shot 2021-11-10 at 12 53 35 PM" src="https://user-images.githubusercontent.com/19869374/141191810-de03833c-f244-4a7f-8b44-62d9a3ad2661.png">

And here's my optimized-Alpha fit result. Doesn't seem consistent with the tutorial discussion...?
<img width="682" alt="Screen Shot 2021-11-10 at 1 11 11 PM" src="https://user-images.githubusercontent.com/19869374/141194125-61446720-1590-4517-b7bf-540324a1f6d0.png">

dacmess commented:
On tutorial 2-3-how-to-use-cbvcorrector:
The 20-sec fast cadence example worked great. I think that is good.
However, I'm getting an error in the Goodness Metrics and CBVCorrector with other Design Matrices section. I suspect it might be related to my python/package configuration, but I'm not sure. It is on the line:
> cbvcorrector.goodness_metric_scan_plot(cbv_type=cbv_type, cbv_indices=cbv_indices, ext_dm=dm);
This is probably not something you need to diagnose for this pull request. 
The bottom of the error list is:
<img width="1107" alt="Screen Shot 2021-11-10 at 2 17 37 PM" src="https://user-images.githubusercontent.com/19869374/141202847-485b4f7f-f24e-47dd-9e3b-c39466569e9f.png">

jcsmithhere commented:
> -In the intro, you might change "Single-Scale" and "Multi-Scale" to "SingleScale" and "MultiScale" to match the actual python keyword values for cbv_type. 

@dacmess, I added a statement saying that we reference the CBV types using CamelCase, that being the convention used by the missions. 

> -You mention setting extrapolate=False sets the CBV values to 0 and will likely lead to a bad fit, but you might caution about extrapolate=True also likely being problematic and maybe it's better to crop the LC (I realize this was in the original code and not just introduced).

I added a sentence warning about the dangers of extrapolation.

jcsmithhere commented:
> More generally, my correction for TIC 99180739 doesn't look like a good correction although the description says that it should look ok. Am I missing something?

@dacmess, I think there might be something wrong with your configuration. You seem to get different results. Your light curve also has additional ungapped cadences. Compare your results with those from the older, published tutorial :

https://docs.lightkurve.org/tutorials/2-creating-light-curves/2-3-how-to-use-cbvcorrector.html 

I wonder if the way your LightKurve configuration is handling quality flags is different than the default. Did you try restarting your kernel and running all cells in the tutorial with no changes?

jcsmithhere commented:
> I reran both tutorials using a poetry install of this branch, rather than adjusting my system path, and I am getting the same behavior and same error as above. Which is on this line in the tutorial:
> % cbvcorrector.goodness_metric_scan_plot(cbv_type=cbv_type, cbv_indices=cbv_indices, ext_dm=dm);
> Cells past this one worked fine, so it appears to just be the scan_plot that is having an issue. I'm still not sure if this is a local setup issue for me, but is poetry supposed to take care of that?

@dacmess, I wonder if it is an incompatibility with your astropy installation. Here is my Lightkurve and astropy install versions:

```
$ conda list | grep lightkurve
lightkurve                2.0.11.dev0               dev_0    <develop>
$ conda list | grep astropy
astropy                   4.2              py38h497a2fe_2    conda-forge
```

dacmess commented:
@jcsmithhere I restarted everything and got similar errors in the 2.3 tutorial. The first difference I noticed from the saved version of the tutorial is on cell #4: cbvCorrector.cbvs[0].plot(); 
My version appear to have only CBVs with any content, whereas the saved version has 16 (see plots below). 
I am running the jupyter notebook through poetry after doing "poetry install" in the lightcurve directory. My various version are:
(base) ➜  lightkurve git:(cbvcorrector_feature/load_20-sec_cbvs) ✗ conda list |grep lightkurve
lightkurve                2.0.6                    pypi_0    pypi
(base) ➜  lightkurve git:(cbvcorrector_feature/load_20-sec_cbvs) ✗ poetry version
lightkurve 2.0.12dev
(base) ➜  lightkurve git:(cbvcorrector_feature/load_20-sec_cbvs) ✗ conda list |grep astropy
astropy                   4.2                      pypi_0    pypi
My conda version of lightkurve is 2.06, by the version running in poetry is 2.0.12dev. Are the two versions stepping on each other, or have I messed up when using poetry?
![caldwell_cbv_plot](https://user-images.githubusercontent.com/19869374/142070971-047a2368-5563-4fb9-bbd4-bc07a18897c7.png)
<img width="843" alt="saved_cbv_plot" src="https://user-images.githubusercontent.com/19869374/142070983-8dd379c7-df00-4dfe-b0ec-dd448518d3dc.png">

jcsmithhere commented:
@dacmess, it does look like you indeed do have two versions of LighKurve stepping on each other. However, it is not clear to me how that would result in you seeing different CBVs in the plot tool than the tutorial. I don't think I changed anything in my code that would result in half the CBVs being zero. It's possible the code you're using is loading only the first 8 CBVs, but I don't think I did anything to change that. 

I really have very little experience with poetry so I can't help there.

dacmess commented:
@jcsmithhere I'll stop spamming this ticket after this, since I think you've got what you need for the 20 sec CBV changes. 

I started from scratch with a new python environment (python 3.8.12) and only installed lightkurve (2.0.12.dev0) and what was needed to run jupyter notebook. Astropy is at version 4.3.1. I am getting the exact same behavior as before; i.e. starting with only 8 valid CBVs. That got me to thinking that the problem might be reprocessed S10 data vs the original? Didn't we switch from delivering all 16 CBVs to not delivering CBVs beyond 8 (or vice versa) with the reprocessing? Could this be the source of the difference between what I'm seeing and the saved tutorial? Of course that doesn't explain the error I'm still getting on the scan_plot when combining CBV and DM corrections.
